### PR TITLE
Feature/custom locales

### LINKED ISSUE #4: Allow the selection for certain locales
### Feature proposal

Based on @mlopezFC 's comment on #3 _"It would be great to be able to only allow the selection for certain locales / countries"_.

What ideas come to your minds about it?

Currently, the addon is limited by the flag-icons package (which provides the flag icons). It renders the flags based on ISO 3166-1-alpha-2 (a 2 letter country code). Because of this, all the supported locales by the addons are:
```java
  public static List<Locale> getAvailableSupportedLocales() {
    return Arrays.stream(Locale.getAvailableLocales()).filter(loc -> loc.getCountry().length() == 2)
        .toList();
  }
```

Maybe it would be a good option to have a helper class or static methods to build only supported locales. Or, we could simply allow the developer to set the locales that they want like:
```java
List<Locale> myDesiredLocales; // list with some locales

LocaleComboBox myCombo = new LocaleComboBox(myDesiredLocales);
// or
myCombo.setItems(myDesiredLocales);
```
The issue with this approach is that it would be possible to add unsupported locales (for which the flags won't show)

So, i'm pretty unsure about it 🤔. Let me know what you think.

### Describe solution expectations

_No response_

### Additional information

_No response_

### LINKED ISSUE #5: Add Helper class or static methods to support more locales
### Feature proposal

It would be nice to have a Helper class used by the LocaleComboBox component with country conversion methods: https://www.iso.org/obp/ui/#search/code
- Numeric to alpha 2
- alpha 3 to alpha 2

### Describe solution expectations

Something like:
```java
public LocaleCountryConverter {

  // all possible conversions based on the iso table (hardcoded maybe?)
  private static Map<String,String> conversions;

  private LocaleCountryConverter() {}
  
    // returns ISO 3166-1-alpha-2 country code
    public static String convert(String countryCode) {
        return conversions.get(countryCode);
    }

}
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by elPeiretti**
- resolves #4 by adding a new constructor that accepts a collection of locales
- resolves #5 by adding the private _getFlagCode_ method and the LocaleContryConverter class

## Summary by CodeRabbit

- **New Features**
  - Enhanced `LocaleComboBox` functionality with a new constructor for customized locale selection.
  - Introduced streamlined retrieval of country codes for improved flag representation.
  - Updated demo component to provide a filtered and sorted list of locales, enhancing user experience.

- **Bug Fixes**
  - Improved logic for displaying country flags by integrating a new method for country code retrieval.

### CODE REVIEW

**Comment by elPeiretti** [comment]
squashed and updated the commits using rebase

File: src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleComboBox.java [hunk 1]
```diff
@@ -143,6 +146,16 @@ private Locale getLocaleForDisplay() {
     }
   }
 
+  private String getFlagCode(Locale locale) {
+    String countryCode = locale.getCountry();
+    Optional<String> isoCode = LocaleCountryConverter.convertToISO3166Code(countryCode);
```
Comment (javier-godoy) [inline src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleComboBox.java]: The following is more idiomatic:
```
return LocaleCountryConverter.convertToISO3166Code(countryCode)
.map(String::toLowerCase)
.orElse(DEFAULT_FLAG_CODE);

File: src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleCountryConverter.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * String alpha2Code = LocaleCountryConverter.convertToISO3166Code("ARG").orElse("Unknown");
+ * </pre>
+ *
+ * @see #convertToISO3166Code(String)
+ */
+public class LocaleCountryConverter {
+
+  private static final int TOTAL_COUNTRIES = 249;
+  private static final List<String> alpha2codes =
+      List.of("AF", "AL", "DZ", "AS", "AD", "AO", "AI", "AQ", "AG", "AR", "AM", "AW", "AU", "AT",
```
Comment (javier-godoy) [inline src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleCountryConverter.java]: That's really hard to maintain. Instead, do

```
  private static final Map<String, String> conversions = new HashMap<>();
  private static void add(String alpha2, String alpha3, int numeric) {
      conversions.put(alpha2, alpha2);
      conversions.put(alpha3, alpha2);
      conversions.put(Integer.toString(numeric), alpha2);
  }
  
  static {
    add("AF","AFG",4);
    //etc ...
  }
```

File: src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleComboBox.java [hunk 3]
```diff
@@ -80,15 +82,22 @@ public enum DisplayMode {
   private Locale customDisplayLocale = Locale.getDefault();
 
   /**
-   * Creates a new instance of LocaleComboBox with all the installed locales.
+   * Creates a new instance of LocaleComboBox.
```
Comment (javier-godoy) [inline src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleComboBox.java]: IMHO `Creates a new instance of {@code LocaleComboBox}.` (but we haven't established a convention for it yet). See `https://github.com/FlowingCode/DevelopmentConventions/issues/40`

**Review by javier-godoy — CHANGES_REQUESTED** [review]

File: src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleComboBox.java (hunk 1 above)
Comment (elPeiretti) [reply src/main/java/com/flowingcode/vaadin/addons/localecombobox/LocaleComboBox.java:160]: I agree with using .orElse(), but the country code shouldn’t be converted to lowercase, as that would deviate from the ISO format

**Review by javier-godoy — APPROVED** [review]

**Pull request merged**